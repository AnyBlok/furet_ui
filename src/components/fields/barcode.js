/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import {defineComponent} from '../factory'
import {fields} from './fields';
import VueBarcode from 'vue-barcode';

defineComponent('furet-ui-list-field-barcode', {
  template:`
    <span v-if="isHidden" />
    <barcode v-else v-bind:value="value" v-bind="config.options">
      Can't generate the barcode
    </barcode>
  `,
  extend: ['furet-ui-list-field-common'],
  prototype: {
    components: {
      barcode: VueBarcode
    },
  },
})
fields.list.barcode = 'furet-ui-list-field-barcode'


defineComponent('furet-ui-form-field-barcode', {
  template: `
    <furet-ui-form-field-common-tooltip-field
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:config="config"
    >
      <barcode v-if="isReadonly" v-bind:value="value" v-bind="config.options">
        Can't generate the barcode
      </barcode>
      <b-input 
        v-else
        v-bind:value="value" 
        v-on:input="updateValue"
        v-bind:maxlength="config.maxlength"
        v-bind:placeholder="config.placeholder"
        icon-pack="fa"
        v-bind:icon="config.icon"
        v-bind:key="config.key"
      />
    </furet-ui-form-field-common-tooltip-field>
  `,
  extend: ['furet-ui-form-field-common'],
  prototype: {
    components: {
      barcode: VueBarcode
    },
  },
})
fields.form.barcode = 'furet-ui-form-field-barcode'
