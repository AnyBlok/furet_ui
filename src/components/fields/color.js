/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import {defineComponent} from '../factory'
import {fields} from './fields';

defineComponent('furet-ui-list-field-color', {
  template:`
    <span v-if="isHidden" />
    <component 
      v-else-if="config.slot" 
      v-bind:is="component_template"
      v-bind:config="config"
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:value="value"
    />
    <b-input 
      v-else-if="value"
      type="color"
      v-bind:value="value" 
      disabled
    />
    <span v-else />
  `,
  extend: ['furet-ui-list-field-common'],
})
fields.list.color = 'furet-ui-list-field-color'

defineComponent('furet-ui-thumbnail-field-color', {
  template: `
    <furet-ui-thumbnail-field-common-tooltip-field
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:config="config"
    >
      <component 
        v-if="config.slot" 
        v-bind:is="component_template"
        v-bind:config="config"
        v-bind:resource="resource"
        v-bind:data="data"
        v-bind:value="value"
      />
      <b-input 
        v-else-if="value"
        type="color"
        v-bind:value="value" 
        disabled
      />
    </furet-ui-thumbnail-field-common-tooltip-field>
  `,
  extend: ['furet-ui-thumbnail-field-common'],
})
fields.thumbnail.color = 'furet-ui-thumbnail-field-color'
fields.kanban.color = 'furet-ui-thumbnail-field-color'

defineComponent('furet-ui-form-field-color', {
  template: `
    <furet-ui-form-field-common-tooltip-field
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:config="config"
    >
      <div class="field has-addons">
        <p v-if="isReadonly" class="control is-expanded">
          <component 
            v-if="config.slot" 
            v-bind:is="component_template"
            v-bind:config="config"
            v-bind:resource="resource"
            v-bind:data="data"
            v-bind:value="value"
          />
          <b-input 
            v-else-if="data"
            type="color"
            v-bind:value="value" 
            disabled
          >
          </b-input>
        </p>
        <p v-else class="control">
          <b-input 
            style="width: 100px"
            v-if="data || !isReadonly"
            type="color"
            v-bind:value="value" 
            v-on:input="updateValue"
            v-bind:disabled="isReadonly"
          >
          </b-input>
        </p>
        <p class="control is-expanded">
          <b-input 
            v-if="!isReadonly"
            v-bind:value="value" 
            v-on:input="updateValue"
            v-bind:placeholder="config.placeholder"
            icon-pack="fa"
            v-bind:icon="config.icon"
            pattern="#[A-Fa-f1-9]{6}"
          />
        </p>
      </div>
    </furet-ui-form-field-common-tooltip-field>
  `,
  extend: ['furet-ui-form-field-common'],
})
fields.form.color = 'furet-ui-form-field-color'
