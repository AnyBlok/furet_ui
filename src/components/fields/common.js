/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import {defineComponent} from '../factory'
import { debounce } from 'debounce';


/**
 * Eval a value (that could be anything String / integer / null / undefined / bool)
 * and cast as a boolean or null.
 *
 * @param {any} value - A value to cast to boolean
 * @returns {Boolean?} return one of those values [true, false, null]
 *
 * Have a look to tests/unit/fields/boolean.spec.js for a complet list of
 * use case.
 */
export const safe_eval_boolean = value => {
  if (value === null || value === undefined) {
    return null;
  }

  // if dev use new String("") or new Boolean(true) this give an object so
  // we convert to its primitive type
  if (
    typeof value === "object" &&
    ["[object String]", "[object Boolean]"].includes(toString.call(value))
  ) {
    value = value.valueOf();
  }

  if (typeof value === "string") {
    if (
      value.trim().toLowerCase() === "null" ||
      value.trim().toLowerCase() === "undefined"
    ) {
      return null;
    }
    return ["true", "t", "yes", "y", "on", "1"].includes(
      value.trim().toLowerCase()
    );
  }

  if (typeof value === "number") {
    return value === 1;
  }

  if (typeof value === "boolean") {
    return value;
  }

  return false;
};

export const safe_eval = (condition, fields, resource) => {
    const now = Date.now(),
          toDate = (v) => new Date(v),
          selectors = Object.assign((resource.manager || {}).selectors || {}, resource.selectors || {}),
          tabs = resource.tabs;
    fields  // lint
    selectors  // lint
    tabs  // lint
    now  // lint
    toDate  // lint
    let res = false;
    try {
        res = eval(condition) ? true : false;
    } catch (e) {
      console.log(e)
    }
    return res;
}

export const listTemplate = `
  <div>
    <span v-if="isHidden" />
    <span v-else>{{value}}</span>
  </div>
`


defineComponent('furet-ui-list-field-common', {
  prototype: {
    props: ['resource', 'data', 'config'],
    computed: {
      value () {
        return this.data[this.config.name] || '';
      },
      isHidden () {
        if (this.config.hidden === undefined) return false;
        return safe_eval(this.config.hidden, this.data || {}, this.resource);
      },
    },
  }
});


defineComponent('furet-ui-form-field-common-tooltip', {
  template: `
    <div v-if="isHidden" />
    <b-tooltip 
      v-bind:label="getTooltip" 
      v-bind:position="tooltipPosition"
      v-bind:style="{'width': '100%'}"
      v-else
    >
      <slot />
    </b-tooltip>`,
  prototype: {
    props: ['resource', 'data', 'config'],
    computed: {
      isHidden () {
        if (this.config.hidden === undefined) return false;
        return safe_eval(this.config.hidden, this.data || {}, this.resource);
      },
      getTooltip () {
        return this.config.tooltip || '';
      },
      tooltipPosition () {
        return this.config.tooltip_position || 'is-top';
      },
    },
  },
})


defineComponent('furet-ui-form-field-common-tooltip-field', {
  template: `
    <furet-ui-form-field-common-tooltip
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:config="config"
    >
      <b-field 
        v-bind:label="config.label"
        v-bind:type="getType"
        v-bind:message="getMessage"
        v-bind:style="{'width': 'inherit'}"
      >
        <slot />
      </b-field>
    </furet-ui-form-field-common-tooltip>`,
  prototype: {
    props: ['resource', 'data', 'config'],
    computed: {
      value () {
        return this.data && this.data[this.config.name] || '';
      },
      isRequired () {
        return safe_eval(this.config.required, this.data || {}, this.resource);
      },
      getType () {
        if (this.isRequired) {
          if (this.value) return 'is-info';
          return 'is-danger';
        }
        return '';
      },
      getMessage () {
        if (this.isRequired) {
          if (!this.value) return this.$i18n.t('fields.common.required');
        }
        return ''
      },
    },
  },
})


defineComponent('furet-ui-form-field-common-readonly', {
  prototype: {
    props: ['resource', 'data', 'config'],
    inject: ['partIsReadonly'],
    computed: {
      isReadonly () {
        if (this.resource.readonly) return true;
        if (this.partIsReadonly()) return true;
        const readonlyParams = safe_eval(this.config.readonly, this.data || {}, this.resource);
        if (this.config.writable) {
          const writableParams = safe_eval(this.config.writable, this.data || {}, this.resource);
          return readonlyParams && ! writableParams
        }
        return readonlyParams;
      },
    },
  },
})


defineComponent('furet-ui-form-field-common', {
  extend: ['furet-ui-form-field-common-readonly'],
  prototype: {
    inject: ['updateChangeState'],
    computed: {
      value () {
        return this.data && this.data[this.config.name] || null;
      },
    },
    methods: {
      updateValue: debounce(function(value, merge) {
        const action = {
          model: this.resource.model,
          pk: this.resource.pks,
          uuid: this.resource.uuid,
          fieldname: this.config.name,
          value,
          merge
        }
        this.updateChangeState(action);
      }, 200),
    },
  }
});
