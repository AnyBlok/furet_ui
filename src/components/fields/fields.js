export const fields = {
  list: {},
  form: {},
  thumbnail: {},
  kanban: {},
};
export default {
    fields,
}
