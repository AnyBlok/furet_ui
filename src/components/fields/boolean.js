// /**
// This file is a part of the FuretUI project
// 
//    Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>
// 
// This Source Code Form is subject to the terms of the Mozilla Public License,
// v. 2.0. If a copy of the MPL was not distributed with this file,You can
// obtain one at http://mozilla.org/MPL/2.0/.
// **/
// import _ from 'underscore';
import {defineComponent} from '../factory'
import {fields} from './fields';
import { listTemplate } from './common';


defineComponent('furet-ui-list-field-boolean', {
  template: `
    <div>
      <span v-if="isHidden" />
      <b-checkbox 
        v-else
        v-model="checked"
        disabled
      />
    </div>`,
  extend: ['furet-ui-list-field-common'],
  prototype: {
    computed: {
      checked () {
        return eval(this.value) ? true : false;
      },
    },
  },
})
fields.list.boolean = 'furet-ui-list-field-boolean'


defineComponent('furet-ui-list-field-yesno', {
  template: listTemplate,
  extend: ['furet-ui-list-field-common'],
  prototype: {
    computed: {
      value () {
        const base = 'components.fields.yesno'
        return this.$t(eval(this.data[this.config.name] || '') ? `${base}.yes` : `${base}.no`);
      },
    },
  },
})
fields.list.yesno = 'furet-ui-list-field-yesno'

//export const FieldThumbnailBoolean = Vue.component('furet-ui-thumbnail-field-boolean', {
//    mixins: [ThumbnailMixin],
//    template: `
//        <div v-if="this.isHidden" />
//        <b-tooltip 
//            v-bind:label="getTooltip" 
//            v-bind:position="tooltipPosition"
//            v-else
//        >
//            <b-checkbox 
//                v-model="value" 
//                disabled
//            >
//                {{this.label}}
//            </b-checkbox>
//        </b-tooltip>`,
//    computed: {
//        value () {
//            return eval(this.data[this.name]) ? true : false;
//        },
//    }
//})


defineComponent('furet-ui-form-field-boolean', {
  template: `
    <furet-ui-form-field-common-tooltip
      v-bind:data="data"
      v-bind:config="config"
    >
      <b-checkbox 
        v-model="checked" 
        v-bind:disabled="isReadonly"
      >
        {{ $t(config.label) }}
      </b-checkbox>
    </furet-ui-form-field-common-tooltip>`,
  extend: ['furet-ui-form-field-common'],
  prototype: {
    computed: {
      checked: {
        get () {
          return eval(this.value) ? true : false;
        },
        set (value) {
          this.updateValue (value);
        },
      },
    },
  },
})
fields.form.boolean = 'furet-ui-form-field-boolean'
