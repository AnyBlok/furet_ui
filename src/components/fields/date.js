/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import moment from 'moment-timezone';
import {defineComponent} from '../factory'
import {fields} from './fields';
import { listTemplate, thumbnailTemplate } from './common';
import {i18n} from '../../i18n';


defineComponent('furet-ui-common-field-date', {
  prototype: {
    computed: {
      value () {
        moment.locale(i18n.locale);
        const value = this.data[this.config.name];
        const timezone = this.$store.state.global.userTimeZone;
        if (value) return moment(value).tz(timezone).format('LL');
        return '';
      },
    },
  },
})

defineComponent('furet-ui-list-field-date', {
  template: listTemplate,
  extend: ['furet-ui-list-field-common', 'furet-ui-common-field-date'],
})
fields.list.date = 'furet-ui-list-field-date'

defineComponent('furet-ui-thumbnail-field-date', {
  template: thumbnailTemplate,
  extend: ['furet-ui-thumbnail-field-common', 'furet-ui-common-field-date'],
})
fields.thumbnail.date = 'furet-ui-thumbnail-field-date'

defineComponent('furet-ui-form-field-date', {
  extend: ['furet-ui-form-field-common'],
  template: `
    <furet-ui-form-field-common-tooltip-field
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:config="config"
    >
      <div v-if="isReadonly">
        <component 
          v-if="config.slot" 
          v-bind:is="component_template"
          v-bind:config="config"
          v-bind:resource="resource"
          v-bind:data="data"
          v-bind:value="readonly_value"
        />
        <span v-else>{{ readonly_value }}</span>
      </div>
      <b-datepicker
        v-else
        v-bind:placeholder="config.placeholder"
        v-bind:value="value" 
        v-on:input="updateDateValue"
        v-bind:editable="config.editable"
        v-bind:show-week-number="config.showWeekNumber"
        icon-pack="fa"
        v-bind:icon="config.icon"
        v-bind:key="config.key"
        trap-focus
      />
    </furet-ui-form-field-common-tooltip-field>
  `,
  prototype: {
    computed: {
      readonly_value () {
        moment.locale(i18n.locale);
        const value = this.data[this.config.name];
        const timezone = this.$store.state.global.userTimeZone;
        if (value) return moment(value).tz(timezone).format('LL');
        return '';
      },
      value () {
        moment.locale(i18n.locale);
        const value = this.data[this.config.name];
        if (!value) return null;
        return new Date(Date.parse(value))
      },
    },
    methods: {
      updateDateValue (value) {
        this.updateValue(value.toISOString())
      },
    },
  },
})
fields.form.date = 'furet-ui-form-field-date'
