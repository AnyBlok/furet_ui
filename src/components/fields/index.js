import { fields } from './fields';
import './common';
import './boolean';
import './datetime';
import './integer';
import './password';
import './timestamp';
import './selection';
import './string';

import './relationship';

import { defineComponent } from '../factory';

defineComponent('furet-ui-unknown-field', {
  template : `<div>Unknow Field <strong>{{config.type}}</strong></div>`,
  prototype: {
    props: ['config', 'data'],
  }
})

defineComponent('furet-ui-field', {
  template : `
    <component 
      v-bind:is="component" 
      v-bind:resource="resource" 
      v-bind:config="config" 
      v-bind:data="data"
    />
  `,
  prototype: {
    props: ['resource', 'config', 'data'],
    computed: {
      component () {
        return (fields[this.resource.type] || {})[this.config.type] || 'furet-ui-unknown-field'
      },
    },
  }
})
