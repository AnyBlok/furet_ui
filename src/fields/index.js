/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import React from 'react';
import plugin from '../plugin';
import './string';
import './selection';
import './date';
import './integer';
import './float';
import './url';
import './uuid';
import './sequence';
import './password';
import './color';
import './text';

/**
 * Unknown field, used only if the wanted field is not available
**/
plugin.set(['field'], {Unknown: (props) => {
    return <div>{props.value}</div>
}});

/**
 * return the component for viewType and fieldType
**/
export const getField = (viewType, fieldType, attribs, value) => {
    let field = plugin.get(['field', viewType, fieldType]);
    if (!field) {
        field = plugin.get(['field', 'Unknown']);
        console.log('getField', viewType, fieldType, attribs, value)
    }
    const id = 'view-type-' + viewType + '-field_type-' + fieldType + '-field-name-' + (attribs.name || '');
    return React.createElement(field, Object.assign({id}, attribs, {value}));
}

export default {
    getField,
}
