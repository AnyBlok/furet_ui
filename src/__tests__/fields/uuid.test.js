/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import React from 'react';
import renderer from 'react-test-renderer';
import {getField} from '../../fields';

test('getField for List', () => {
    const component = renderer.create(getField('List', 'UUID', {}, 'Test'));
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('getField for Thumbnail', () => {
    const component = renderer.create(getField('Thumbnail', 'UUID', {}, 'Test'));
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('getField for Form', () => {
    const component = renderer.create(getField('Form', 'UUID', {}, 'Test'));
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
