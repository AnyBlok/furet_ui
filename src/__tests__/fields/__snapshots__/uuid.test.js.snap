// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getField for Form 1`] = `
<div
  className="form-group"
>
  <label
    htmlFor="view-type-Form-field_type-UUID-field-name-"
    style={Object {}}
  >
     
  </label>
  <input
    className="form-control"
    disabled={true}
    id="view-type-Form-field_type-UUID-field-name-"
    onChange={[Function]}
    type="text"
    value="Test"
  />
</div>
`;

exports[`getField for List 1`] = `
<div>
  <span>
    Test
  </span>
</div>
`;

exports[`getField for Thumbnail 1`] = `
<div
  className="form-group"
>
  <label
    htmlFor="view-type-Thumbnail-field_type-UUID-field-name-"
  >
     
  </label>
  <div>
    <span>
      Test
    </span>
  </div>
</div>
`;
